### PR TITLE
[IOTDB-4038] Add the leader metrics to the cluster

### PR DESCRIPTION
**Pull request by 23931017wu**
### After
image
image

### CODE REVIEW

File: confignode/src/main/java/org/apache/iotdb/confignode/manager/load/LoadManager.java [hunk 1]
```diff
@@ -590,7 +591,54 @@ public int getUnknownDataNodesNum() {
...
-  public void addMetrics() {
+  /**
+   * Get the LeaderCount of each DataNodeId
+   *
+   * @return Map<DataNodeId, LeaderCount>
+   */
+  public Map<Integer, Integer> getLeadershipCountByDatanode() {
+    Map<Integer, Integer> idToCountMap = new ConcurrentHashMap<>();
+
+    regionGroupCacheMap.forEach(
```
Comment (Beyyes) [inline confignode/src/main/java/org/apache/iotdb/confignode/manager/load/LoadManager.java:602]: 602-610 can be replaced with `regionGroupCacheMap.forEach(
        (consensusGroupId, regionGroupCache) -> idToCountMap.merge(regionGroupCache.getLeaderDataNodeId(), 1, Integer::sum));`

File: confignode/src/main/java/org/apache/iotdb/confignode/manager/load/LoadManager.java [hunk 2]
```diff
@@ -590,7 +591,54 @@ public int getUnknownDataNodesNum() {
     return allDataNodes.size();
   }
 
-  public void addMetrics() {
+  /**
+   * Get the LeaderCount of each DataNodeId
+   *
+   * @return Map<DataNodeId, LeaderCount>
+   */
+  public Map<Integer, Integer> getLeadershipCountByDatanode() {
```
Comment (Beyyes) [inline confignode/src/main/java/org/apache/iotdb/confignode/manager/load/LoadManager.java:599]: Is this method can be private?

File: confignode/src/main/java/org/apache/iotdb/confignode/manager/load/LoadManager.java [hunk 3]
```diff
@@ -590,7 +591,54 @@ public int getUnknownDataNodesNum() {
...
+  public void addLeaderCount() {
+    Map<Integer, Integer> idToCountMap = getLeadershipCountByDatanode();
+    getNodeManager()
+        .getRegisteredDataNodes(-1)
+        .forEach(
+            dataNodeInfo -> {
+              TDataNodeLocation dataNodeLocation = dataNodeInfo.getLocation();
+              int dataNodeId = dataNodeLocation.getDataNodeId();
+              if (idToCountMap.containsKey(dataNodeId)) {
+                String name =
```
Comment (Beyyes) [inline confignode/src/main/java/org/apache/iotdb/confignode/manager/load/LoadManager.java:623]: I think this line can be extracted to a common method.
It has been used 5 times.

image

**Review by neuyilan — APPROVED** [review]
LGTM

File: confignode/src/main/java/org/apache/iotdb/confignode/manager/load/LoadManager.java [hunk 4]
```diff
@@ -327,7 +327,8 @@ private void updateNodeLoadStatistic() {
       broadcastLatestRegionRouteMap();
     }
     if (nodeCacheMap.size() == getNodeManager().getRegisteredNodeCount()) {
-      addMetrics();
+      addNodeMetrics();
+      addLeaderCount();
```
Comment (Beyyes) [inline confignode/src/main/java/org/apache/iotdb/confignode/manager/load/LoadManager.java:331]: UpdateNodeLoadStatistic method is executed once in a second, do `addNodeMetrics();` and `addLeaderCount();` need to be executed so frequently?

image

File: confignode/src/main/java/org/apache/iotdb/confignode/manager/load/LoadManager.java (hunk 1 above)
Comment (23931017wu) [reply confignode/src/main/java/org/apache/iotdb/confignode/manager/load/LoadManager.java:602]: Okay, learned ~

File: confignode/src/main/java/org/apache/iotdb/confignode/manager/load/LoadManager.java (hunk 3 above)
Comment (23931017wu) [reply confignode/src/main/java/org/apache/iotdb/confignode/manager/load/LoadManager.java:623]: Okay, I'll add it in the tool class.

**Comment by SpriCoder** [comment]
Besides, please update doc of new metrics at the some time.

**Review by SpriCoder — CHANGES_REQUESTED** [review]

File: confignode/src/main/java/org/apache/iotdb/confignode/manager/load/LoadManager.java [hunk 5]
```diff
@@ -590,7 +591,54 @@ public int getUnknownDataNodesNum() {
...
+    Map<Integer, Integer> idToCountMap = new ConcurrentHashMap<>();
+
+    regionGroupCacheMap.forEach(
+        (consensusGroupId, regionGroupCache) -> {
+          Integer count = idToCountMap.get(regionGroupCache.getLeaderDataNodeId());
+          if (count == null) {
+            idToCountMap.put(regionGroupCache.getLeaderDataNodeId(), 1);
+          } else {
+            idToCountMap.put(regionGroupCache.getLeaderDataNodeId(), count + 1);
+          }
```
Comment (CRZbulabula) [inline confignode/src/main/java/org/apache/iotdb/confignode/manager/load/LoadManager.java:609]: Better use getAllLeadership interface to complete these code~

File: confignode/src/main/java/org/apache/iotdb/confignode/manager/load/LoadManager.java [hunk 6]
```diff
@@ -590,7 +591,54 @@ public int getUnknownDataNodesNum() {
...
   }
 
-  public void addMetrics() {
+  /**
+   * Get the LeaderCount of each DataNodeId
+   *
+   * @return Map<DataNodeId, LeaderCount>
+   */
+  public Map<Integer, Integer> getLeadershipCountByDatanode() {
+    Map<Integer, Integer> idToCountMap = new ConcurrentHashMap<>();
```
Comment (CRZbulabula) [inline confignode/src/main/java/org/apache/iotdb/confignode/manager/load/LoadManager.java:600]: Add annotation to this variable like:

`// Map<DataNodeId, leaderCount>`

By the way, you can use Map<Integer, AtomicInteger> and map.computeIfAbsent to simplify the following code logic

**Review by CRZbulabula — CHANGES_REQUESTED** [review]
PTAL~

**Comment by 23931017wu** [comment]
The name attribute that was modified was just the structure of "IP:port", so I removed the outer layer of "EndPoint()".
image
image

**Review by CRZbulabula — APPROVED** [review]
LGTM～

**Review by SpriCoder — APPROVED** [review]
LGTM~

File: confignode/src/main/java/org/apache/iotdb/confignode/manager/load/LoadManagerMetrics.java [hunk 7]
```diff
@@ -0,0 +1,277 @@
...
+
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+/** This class collates metrics about loadManager */
+public class LoadManagerMetrics {
+
+  private final IManager configManager;
+  Map<Integer, Integer> idToCountMap = new ConcurrentHashMap<>();
```
Comment (Beyyes) [inline confignode/src/main/java/org/apache/iotdb/confignode/manager/load/LoadManagerMetrics.java:42]: This variable has not been used.

File: confignode/src/main/java/org/apache/iotdb/confignode/manager/load/LoadManagerMetrics.java [hunk 8]
```diff
@@ -0,0 +1,277 @@
...
+              NodeStatus.Unknown.toString());
+    }
+  }
+
+  /**
+   * Get the LeaderCount of each DataNodeId
+   *
+   * @return Map<DataNodeId, LeaderCount>
+   */
+  public Integer getLeadershipCountByDatanode(int DataNodeId) {
```
Comment (Beyyes) [inline confignode/src/main/java/org/apache/iotdb/confignode/manager/load/LoadManagerMetrics.java:203]:
```suggestion
  public Integer getLeadershipCountByDatanode(int dataNodeId) {
```

File: confignode/src/main/java/org/apache/iotdb/confignode/manager/load/LoadManagerMetrics.java [hunk 9]
```diff
@@ -0,0 +1,277 @@
...
+        .getMetricManager()
+        .removeGauge(
+            Metric.DATA_NODE.toString(),
+            Tag.NAME.toString(),
+            "total",
+            Tag.STATUS.toString(),
+            NodeStatus.Unknown.toString());
+  }
+
+  private LoadManager getLoadManager() {
```
Comment (Beyyes) [inline confignode/src/main/java/org/apache/iotdb/confignode/manager/load/LoadManagerMetrics.java:270]: never used

File: registerd [hunk 10]
```diff
@@ -0,0 +1,2 @@
+[IOTDB-3955 37b6079ef3] registed->G:\deskfile\work\BONC\IoTDB\IoTDB_db\iotdb01\iotdb>git commit -m registed-
```
Comment (Beyyes) [inline registerd:1]: [important] remove this file

**Review by Beyyes — APPROVED** [review]

**Pull request merged**